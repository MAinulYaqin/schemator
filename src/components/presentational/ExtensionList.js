import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

import Extension from "./Extension";

const Container = styled.div`
  width: 240px;
  height: 100%;
  overflow-y: auto;
  background: #333;
`;

const EmptyText = styled.p`
  font-size: 12px;
  color: white;
  font-family: Roboto;
  text-align: center;
  padding: 16px;
  margin: 0px;
`;

const ExtensionList = ({ items, keyword, active, onClickItem }) => {
  const byName = item => item.name.match(new RegExp(keyword, "i"));
  const makeExcerpt = item => {
    const suffix = item.description.length > 22 ? "..." : "";
    const excerpt = item.description.slice(0, 22);
    const description = `${excerpt}${suffix}`;

    return { ...item, description };
  };
  const filteredItems = items.filter(byName).map(makeExcerpt);

  return (
    <Container>
      {filteredItems.length ? (
        filteredItems.map((item, index) => (
          <Extension
            key={index}
            {...item}
            onClick={() => onClickItem(item)}
            isActive={active === item.id}
          />
        ))
      ) : (
        <EmptyText>No extensions found</EmptyText>
      )}
    </Container>
  );
};

ExtensionList.propTypes = {
  items: PropTypes.array,
  keyword: PropTypes.string,
  active: PropTypes.string,
  onClickItem: PropTypes.func
};

ExtensionList.defaultProps = {
  items: [],
  keyword: "",
  onClickItem: () => {}
};

export default ExtensionList;
