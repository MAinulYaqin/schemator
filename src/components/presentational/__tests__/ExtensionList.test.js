import React from "react";
import { shallow } from "enzyme";

import ExtensionList from "../ExtensionList";

describe("ExtensionList", () => {
  const setup = propOverrides => {
    const props = {
      items: [
        {
          name: "Laravel Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export to Laravel file"
        },
        {
          name: "SQL Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export to SQL file"
        },
        {
          name: "MongoDB Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export to MongoDB file"
        }
      ],
      onClickItem: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<ExtensionList {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders with excerpt", () => {
    const { wrapper } = setup({
      items: [
        {
          name: "Laravel Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export scheme to Laravel file"
        },
        {
          name: "SQL Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export scheme to SQL file"
        },
        {
          name: "MongoDB Exporter",
          image: "image.png",
          author: "Gattigaga Hayyuta Dewa",
          description: "Export scheme to MongoDB file"
        }
      ]
    });

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders with keyword", () => {
    const { wrapper } = setup({ keyword: "mongo" });

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders without extensions", () => {
    const { wrapper } = setup({ items: [] });

    expect(wrapper).toMatchSnapshot();
  });

  it("should calls 'onClickItem' while item was clicked", () => {
    const { wrapper, props } = setup();

    props.items.forEach((item, index) => {
      wrapper
        .find("Extension")
        .at(index)
        .simulate("click", item);

      expect(props.onClickItem).toBeCalledWith(item);
    });
  });
});
