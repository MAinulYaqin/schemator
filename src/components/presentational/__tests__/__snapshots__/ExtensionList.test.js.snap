// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtensionList should renders default 1`] = `
<styled.div>
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export to Laravel file"
    image="image.png"
    key="0"
    name="Laravel Exporter"
    onClick={[Function]}
  />
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export to SQL file"
    image="image.png"
    key="1"
    name="SQL Exporter"
    onClick={[Function]}
  />
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export to MongoDB file"
    image="image.png"
    key="2"
    name="MongoDB Exporter"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`ExtensionList should renders with excerpt 1`] = `
<styled.div>
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export scheme to Larav..."
    image="image.png"
    key="0"
    name="Laravel Exporter"
    onClick={[Function]}
  />
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export scheme to SQL f..."
    image="image.png"
    key="1"
    name="SQL Exporter"
    onClick={[Function]}
  />
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export scheme to Mongo..."
    image="image.png"
    key="2"
    name="MongoDB Exporter"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`ExtensionList should renders with keyword 1`] = `
<styled.div>
  <Extension
    author="Gattigaga Hayyuta Dewa"
    description="Export to MongoDB file"
    image="image.png"
    key="0"
    name="MongoDB Exporter"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`ExtensionList should renders without extensions 1`] = `
<styled.div>
  <styled.p>
    No extensions found
  </styled.p>
</styled.div>
`;
